EndpointSlice API GA Doc Updates for 1.21 (#27070)
kubernetes/website: content/en/docs/concepts/services-networking/endpoint-slices.md

### Topology information {#topology}

Each endpoint within an EndpointSlice can contain relevant topology information.
This is used to indicate where an endpoint is, containing information about the

> corresponding Node, zone. When the values are available, the

Reviewer: nit: Possibly reword?
Each endpoint within an EndpointSlice can contain relevant topology information.
The topology information includes the location of the endpoint and information about the corresponding Node and zone.

Author: Thanks for the suggestion!

control plane sets the following per endpoint fields for EndpointSlices:

* `nodeName` - The name of the Node this endpoint is on.
* `zone` - The zone this endpoint is in.

#### Deprecated per endpoint `topology` field

> In the GA API, the field `deprecatedTopology` was added for backwards

Reviewer: nit: Could add a new line between the heading and the first sentence.

compatibility with the `topology` field that exists on the v1beta1 API. The

> `topology` will be converted to `depcreatedTopology` and will be available as

Reviewer: When will this conversion happen?

Author: I have updated this section to be a little more general. I realized the conversion probably goes into too much detail of the actual implementation. Please let me know if you think the new version is clearer or if it does not need more detail.

a read only field for `v1` clients. Any attempts by a `v1` client to write to

> `deprecatedTopology` will cause the entire field to dropped. The

Reviewer: Reword:
Any attempts by a `v1` client to write to `deprecatedTopology` causes the entire field to be ignored?

### Topology information {#topology}

Each endpoint within an EndpointSlice can contain relevant topology information.
The topology information includes the location of the endpoint and information
about the corresponding Node and zone. This will be available in the following
per endpoint fields on EndpointSlices:

* `nodeName` - The name of the Node this endpoint is on.
* `zone` - The zone this endpoint is in.

{{< note >}}

> In the GA API, the per endpoint `topology` was effectively removed in favor of

Reviewer 2: Tiny Nit: I think "v1 API" might be slightly easier to understand, but I'm not sure if we have a standard here for docs.

Author: I will go with "v1 Api" for now then. I wasn't sure what the standard was here either.

the dedicated fields `nodeName` and `zone`.

If the only keys used in the `topology` field were
`kubernetes.io/hostname` and `topology.kubernetes.io/zone`, those values
will be available in the per endpoint `nodeName` and `zone` fields
respectively. Any other keys will require continued usage of the

> `v1beta1` API.

Reviewer 2: This is a confusing thing to document, and I'm not sure how much detail is appropriate here. Maybe something along the lines of this would work:

> Setting arbitrary topology fields per endpoint has been deprecated and will not be supported in the v1 API. Instead, the v1 API supports setting individual node name and zone fields. These fields will be automatically translated between API versions. For example, the value of `topology[topology.kubernetes.io/zone]` in the v1beta1 API will be accessible in the `zone` field in the v1 API.

Author: That sounds much clearer, thanks!

### Topology information {#topology}

Each endpoint within an EndpointSlice can contain relevant topology information.
The topology information includes the location of the endpoint and information

> about the corresponding Node and zone. This will be available in the following

Reviewer 3: ```suggestion
about the corresponding Node and zone. These are available in the following
```

per endpoint fields on EndpointSlices:

* `nodeName` - The name of the Node this endpoint is on.
* `zone` - The zone this endpoint is in.

{{< note >}}
In the v1 API, the per endpoint `topology` was effectively removed in favor of
the dedicated fields `nodeName` and `zone`.

> Setting arbitrary topology fields per endpoint has been deprecated and will not

Reviewer 3: ```suggestion
Setting arbitrary topology fields per Endpoints has been deprecated and is not supported
```

Reviewer 2: Agree that "endpoint" makes more sense in this case. (The naming of these two resources is rather confusing).

Reviewer 3: I was suggesting using `Endppoints` because I am not sure if there is such resource called `endpoint`. The clarification makes sense. However, if the `endpoint` in this context means a field in the `EndpointSlice`, we may want to add backticks around 'endpoint` to make it clearer or more explicitly like the following:

> Setting arbitrary topology fields on the `endpoint` field of a `EndpointSlice` resource has been deprecated.

be supported in the v1 API. Instead, the v1 API supports setting individual

> node name and zone fields. These fields will be automatically translated

Reviewer 3: ```suggestion
`nodeName` and `zone` fields. These fields are automatically translated
```

between API versions. For example, the value of

> topology[topology.kubernetes.io/zone] in the v1beta1 API will be accessible in

Reviewer 3: ```suggestion
`topology.kubernetes.io/zone` in the v1beta1 API are available as
```

Author: The field in the API is `topology` which is a map. In this case we are referring to the key value pair `topology[topology.kubernetes.io/zone]`. I added backticks, does that make it clearer?

Reviewer 2: Maybe `topology["topology.kubernetes.io/zone"]` would be slightly clearer? Agree that the `topology[` prefix adds important context here.

Reviewer 3: I see. Since `topology` is a map, and `topology.kubernetes.io/zone` is a key in that map, we may want to revise the sentence so that the text is more clearer for YAML developers/users.

Author: I have called out that `topology.kubernetes.io/zone` as a key in the `topology` field in the new version. Let me know if the text is clear enough now. I am not exactly sure how to describe it for YAML users.

> the zone field in the v1 API.

Reviewer 3: ```suggestion
the `zone` field in the v1 API.
```
